_posts/2020-07-16-building-drivers-on-zephyr.md

```diff
@@ -0,0 +1,644 @@
+
+These "runners" are set in the board configuration. Here's the `board.cmake` for the nRF9160 Feather:
+
+```
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_args(nrfjprog "--nrf-family=NRF91")
+board_runner_args(jlink "--device=cortex-m33" "--speed=4000")
+include(${ZEPHYR_BASE}/boards/common/nrfjprog.board.cmake)
+include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
+```
+
+As you can see, the nRF9160 Feather supports the `nrfjprog` and `jlink` options.
+
+## Dependency Management
+
+There are a few different ways of handling dependencies in firmware land. I've personally used Git submodules as my dependency "manager" of choice. Zephyr though, is very opinionated on how you should work with your dependencies. Let's look into how we do that with `west` and a `west.yml` file.
+
+`west.yml` is a manifest file much like a `Cargo.toml` in rust or `package.json` for Node. When you initialize a Zephyr repository using `west init`, `west` will recursively shallow clone the dependencies and put them in place.
+
+Here's what a `west.yml` file may look like. This was for a personal project that uses [NanoPB](https://github.com/nanopb/nanopb) and also the nRF Connect SDK.
+
+```yaml
+manifest:
+  remotes:
+    - name: nrfconnect
+      url-base: https://github.com/nrfconnect
+    - name: nanopb
+      url-base: https://github.com/nanopb
+  projects:
+    - name: nrf
+      remote: nrfconnect
+      revision: v1.3.0
+      import: true
+    - name: nanopb
+      repo-path: nanopb
+      revision: 0.3.9
+      path: pyrinas/ext/nanopb
+      remote: nanopb
+  self:
+    # This repository should be cloned to ncs/nrf.
+    path: pyrinas
+```
+
+The `west.yml` file doesn't directly affect driver operation but if you're going to include your drivers as a separate repo, this is where you would add it! You can learn more though by going to the Zephyr documentation on [West Manifests.](https://docs.zephyrproject.org/1.14.1/guides/west/manifest.html) (complete with crazy dependency diagrams and more!)
+
+## How Things Compile
+
+One of the important things to know about driver development is how things compile. It may bend your mind a bit but we'll get through this together.
+
+### `CMakeLists.txt`
+
+If you look carefully, many folders and sub-folders in Zephyr have a consistent pair of files: `CMakeLists.txt` and `Kconfig`. These files is where all the magic happens.
+
+You may already have an idea of what `CMakeLists.txt` does by its name. It's a Makefile of sorts that is used to add source files, folders and more to a project. The neat thing about `cmake` is that it's recursive (for better or for worse) and higher level `CMakeLists.txt` files dictate what, if anything, is below it. There's no more ginormous Makefiles. You make the changes you need closest to the code that's affected.
+
+Sometimes `CMakeLists.txt` are used to simply "include" more directories like in this example:
+
+```
```

Comment:
Should be able to highlight CMake with using

"```cmake"


at the top. For the list of other options, you can check here: http://rouge.jneen.net/

Reply:
Fixed! I didn't see one for Kconfig though. Bummer.